typing/oprint.ml

```diff
@@ -459,38 +459,69 @@ let out_sig_item = ref (fun _ -> failwith "Oprint.out_sig_item")
 let out_signature = ref (fun _ -> failwith "Oprint.out_signature")
 let out_type_extension = ref (fun _ -> failwith "Oprint.out_type_extension")
 
-let rec print_out_functor funct ppf =
-  function
-    Omty_functor (_, None, mty_res) ->
-      if funct then fprintf ppf "() %a" (print_out_functor true) mty_res
-      else fprintf ppf "functor@ () %a" (print_out_functor true) mty_res
-  | Omty_functor (name, Some mty_arg, mty_res) -> begin
-      match name, funct with
-      | "_", true ->
-          fprintf ppf "->@ %a ->@ %a"
-            print_simple_out_module_type mty_arg
-            (print_out_functor false) mty_res
-      | "_", false ->
+let rec print_out_module_type ppf mty =
+  print_out_functor ppf mty
+and print_out_functor ppf = function
+  | Omty_functor _ as t ->
+     (* For anonymous functor arguments, the logic to choose between
+        the long-form
+          functor (_ : S) -> ...
+        and the short-form
+          S -> ...
+        is as follows: if we are already printing long-form functor arguments,
+        we use the long form unless all remaining functor arguments can use
+        the short form. (Otherwise use the short form.)
+
+        For example,
+          functor (X : S1) (_ : S2) (Y : S3) (_ : S4) (_ : S5) -> sig end
+        will get printed as
+          functor (X : S1) (_ : S2) (Y : S3) -> S4 -> S5 -> sig end
+
+        but
+          functor (_ : S1) (_ : S2) (Y : S3) (_ : S4) (_ : S5) -> sig end
+        gets printed as
+          S1 -> S2 -> functor (Y : S3) -> S4 -> S5 -> sig end
+      *)
+     let rec print_functor ppf = function
+       | Omty_functor ("_", Some mty_arg, mty_res) ->
           fprintf ppf "%a ->@ %a"
             print_simple_out_module_type mty_arg
-            (print_out_functor false) mty_res
-      | name, true ->
-          fprintf ppf "(%s : %a)@ %a" name
-            print_out_module_type mty_arg (print_out_functor true) mty_res
-      | name, false ->
-            fprintf ppf "functor@ (%s : %a)@ %a" name
-              print_out_module_type mty_arg (print_out_functor true) mty_res
-    end
-  | m ->
-      if funct then fprintf ppf "->@ %a" print_out_module_type m
-      else print_out_module_type ppf m
-and print_out_module_type ppf = function
-  | Omty_functor _ as t -> fprintf ppf "@[<2>%a@]" (print_out_functor false) t
+            print_functor mty_res
+       | Omty_functor _ as non_anonymous_functor ->
+          let rec collect_args acc = function
+            | Omty_functor (name, mty_arg, mty_res) ->
+               collect_args ((name, mty_arg) :: acc) mty_res
+            | non_functor -> uncollect_anonymous_suffix acc non_functor
```

Comment:
I have one last small nitpicking remark: `collect_args` and `uncollect_anonymous_suffix` seems to be mutually recursive by accident. Indeed, if `collect_args` stopped at
```OCaml
| non_functor -> acc, non_functor
```
it would be the very natural functions that would map the inline-style list of arguments followed by the result type representation to the non-inline reverse list of arguments and result type one. I think it may even make sense to name this function outside of `print_out_functor` in this case.

Reply:
(This was fixed and rebased.)